R/client.R

```diff
@@ -0,0 +1,90 @@
+#' Create seaweed client, to manage HTTP requests to SeaweedFS
+#'
+#' @keywords internal
+#' @noRd
+seaweed_client <- R6::R6Class(
+  "seaweed_client",
+  cloneable = FALSE,
+
+  public = list(
+    seaweed_url = NULL,
+
+    #' @description
+    #' Create client object for sending http requests to seaweed
+    #'
+    #' @param seaweed_url Root URL of Seaweed master
+    #'
+    #' @return A new `seaweed_client` object
+    initialize = function(seaweed_url) {
+      self$seaweed_url <- seaweed_url
+    },
+
+    #' @description
+    #' Send GET request to SeaweedFS master
+    #'
+    #' @param ... Args passed on to httr
+    #'
+    #' @return Response from SeaweedFS
+    GET = function(...) {
+      self$request(httr::GET, ...)
+    },
+
+    #' @description
+    #' Send POST request to SeaweedFS master
+    #'
+    #' @param ... Args passed on to httr
+    #'
+    #' @return Response from SeaweedFS
+    POST = function(...) {
+      self$request(httr::POST, ...)
+    },
+
+    #' @description
+    #' Send DELETE request to SeaweedFS
+    #'
+    #' @param ... Args passed on to httr
+    #'
+    #' @return Response from SeaweedFS
+    DELETE = function(...) {
+      self$request(httr::DELETE, ...)
+    },
+
+    #' @description
+    #' Send an HTTP request to SeaweedFS master
+    #'
+    #' @param verb A httr function for type of request to send e.g. httr::GET
+    #' @param path Request path
+    #' @param ... Additional args passed on to httr
+    #'
+    #' @return Response from SeaweedFS
+    request = function(verb, path, ...) {
+      url <- paste0(self$seaweed_url, "/", path)
+      res <- verb(url, scheme = "http", ...)
+      self$parse_response(res)
+    },
+
+    parse_response = function(res) {
+      code <- httr::status_code(res)
+      if (code >= 400 && code < 600) {
+        stop(kelp_error(res))
+      }
+      ## Can be JSON or plain text response
+      httr::content(res, encoding = "UTF-8")
```

Comment:
I would recommend making the json-ness of this be an input parameter to this function rather than detecting it (we do that elsewhere, I can dig it up if needed)

Reply:
Actually I think this is no longer the case, since adding separate error handling above. When an error seaweedFS returns JSON. When it is successful we get plain text. I've removed the comment.

Reply:
OK, in that case this should be

```suggestion
      httr::content(res, "text", encoding = "UTF-8")
```

but this is not critical

Reply:
Fixed this properly now, we set it explicitly. Above comment is nonsense, this returns JSON most of the time and returns plain text when reading a file